Solving issue #1065. Playnite now checks every available drive for th…

MopsiMauser opened the pull request:
…e given path for Images/ROMs, emulator executables and manually added games if the respective path does not exist by using the given path and just changing the drive when a game is about to be started

I have verified that:
- [x ] These changes work, by building the application and testing them.
- [x ] Pull request is targeting `devel` branch.
- [ ] I added myself into [contributors file](https://github.com/JosefNemec/Playnite/blob/devel/source/Playnite.DesktopApp/Resources/contributors.txt) if I want to receive contribution credit in the application itself.

JosefNemec commented on source/Playnite/Controllers/GenericGameController.cs:
```diff
@@ -133,5 +145,59 @@ private void Monitor_TreeDestroyed(object sender, EventArgs args)
+
+        private void CheckEmulatorConfigExecutable(EmulatorProfile emulatorProfile)
+        {
+            if (Game.PlayAction.Type == GameActionType.Emulator)
```
Working directory has to be handled as well. Some Play actions have relative Path, but use absolute WorkingDir instead. The same goes for emulators.

MopsiMauser commented on source/Playnite/Controllers/GenericGameController.cs:
```diff
@@ -133,5 +145,59 @@ private void Monitor_TreeDestroyed(object sender, EventArgs args)
             OnStopped(this, new GameControllerEventArgs(this, stopWatch.Elapsed.TotalSeconds));
         }
+
+        private void CheckGameImagePath()
```
So these specific check methods should just return the path which is then assigned to the Game object in the Play method? I think i need to change the Game object because it is directly used to expand the variables as far as i understand it.

JosefNemec replied:
You can create copy (for example via `GetClone` extension method) and modify that copy. I just don't like the idea of modify the input object since it's probably used elsewhere up the call stack.

MopsiMauser commented:
Thank you for your advices. I refactored the code and added the same logic when an installation directory is about to be opened via context menu. @JosefNemec

JosefNemec commented on source/Playnite.Common/FileSystem.cs:
```diff
@@ -325,5 +325,46 @@ public static void CopyDirectory(string sourceDirName, string destDirName, bool
+        {
+            if (!Paths.IsFullPath(originalPath))
+            {
+                if (errorIfNotFullPath)
```
I really think we don't need to have an option to throw an error. Especially if we don't use it anywhere.

JosefNemec commented on source/Playnite/Controllers/GenericGameController.cs:
```diff
@@ -20,26 +21,43 @@ public class GenericGameController : BaseGameController
+            var emulators = database.Emulators.ToList();
+            var profileClone = GameActionActivator.GetGameActionEmulatorConfig(playAction, emulators)?.GetClone();
+
+            //Solving Issue #1065 if removable drives are used for storing ROMs/Images
```
Can you remove references to #1065? I don't think it's necessary. This is not workaround for some issue, but new feature and I usually don't add references for those.

JosefNemec commented on source/Playnite/Controllers/GenericGameController.cs:
```diff
@@ -20,26 +21,43 @@ public class GenericGameController : BaseGameController
-            var playAction = Game.PlayAction.ExpandVariables(Game);
+            var gameClone = Game.GetClone();
+            var playAction = gameClone.PlayAction;
+            var emulators = database.Emulators.ToList();
```
Why are you creating copy list of emulators? I don't see it use anywhere, but in one place, and I don't see a reason for this. Please correct me if I'm wrong.

MopsiMauser replied:
I did not add ToList() call. It was already there. I guess because the GameActionActivator.GetGameActionEmulatorConfig(playAction, emulators) method needs an List<Emulator> instance as second parameter

JosefNemec replied:
Hmm, ok then.

JosefNemec commented on source/Playnite/Controllers/GenericGameController.cs:
```diff
@@ -20,26 +21,43 @@ public class GenericGameController : BaseGameController
 
             Dispose();
+
+            profileClone = profileClone?.ExpandVariables(gameClone);
```
Can you move it above dispose call? It look weird just separated here.

JosefNemec commented on source/Playnite/GamesEditor.cs:
```diff
@@ -260,7 +260,19 @@ public void OpenGameLocation(Game game)
             try
             {
-                Process.Start(game.ExpandVariables(game.InstallDirectory));
+                var gameClone = game.GetClone();
```
Correct if I'm wrong, but I don't see a reason to create game copy here.

MopsiMauser replied:
I am creating a copy, because I change the provided Game object. But as you mention it now, i just noticed that InstallDirectory can contain placeholder and so i need to resolve these placeholders before I can check if the path or alternatives exist

JosefNemec replied:
Just store original install directory into a variable and work with that, making copy of the game object is unnecessary.

MopsiMauser replied:
But first I need to expand the placeholders before I check for alternatives, don't I?

MopsiMauser replied:
But I think it makes actually no sense using placeholders for install direcotry because it is used to resolve placeholder {InstallDir}, which could lead to the problem that {InstallDir} is resolved with the value {InstallDir}, or am I wrong? ;-)

JosefNemec replied:
I don't know why there's expanding going on in the first place. I think we can remove that, because it IMO doesn't make sense to have variables in installation directory (I can't think of a use case for them there). I don't remember why I added expanding here.

JosefNemec commented on source/Playnite.Common/FileSystem.cs:
```diff
@@ -325,5 +325,46 @@ public static void CopyDirectory(string sourceDirName, string destDirName, bool
+                }
+                else
+                {
+                    return originalPath;
```
Should return empty string as well.

JosefNemec commented on source/Playnite.Common/FileSystem.cs:
```diff
@@ -325,5 +325,46 @@ public static void CopyDirectory(string sourceDirName, string destDirName, bool
+            return CheckDrivesForPath(directoryPath, errorIfNotFullPath, path => Directory.Exists(path));
+        }
+
+        private static string CheckDrivesForPath(string originalPath, bool errorIfNotFullPath, Predicate<string> predicate)
```
Can you try/catch the whole method just to be sure? log the error, of course, but also add condition to catch only if debugger is not attached. Thanks

I don't think this should generally fail, but I've seen of lot of weird stuff happening on people's PCs, so just to be sure. Especially if this is just a bonus feature for cases where users use wrong paths in the first place.

MopsiMauser replied:
Do you mean this check when (!PlayniteEnvironment.ThrowAllErrors) or actually !Debugger.IsAttached ?

JosefNemec replied:
!Debugger.IsAttached

MopsiMauser commented:
I just tested it using UNC path and I discovered a bug. When expanding the path for a GameAction Paths.FixSeparators is called which replaces \\ with \ and an invaild path is generated, "\\LAPTOP-5BNL27DP\KeePass Password Safe 2\Keepass.exe" becomes "\LAPTOP-5BNL27DP\KeePass Password Safe 2\Keepass.exe" (please don't mind that Keepass isn't actually a game. I just added it manually for testing purpose)

JosefNemec commented:
I was talking about mounted UNC paths, when you connect it as new drive. There are known issues when using direct UNC paths, ignore those. There's a separate issue to fix them.

MopsiMauser commented:
Ah, ok. I didn't understand it properly. Now I checked it and it works with mounted UNC paths.

MopsiMauser commented:
@JosefNemec I refactored the code you commented and pushed it

JosefNemec commented on source/Playnite.Common/FileSystem.cs:
```diff
@@ -325,5 +326,48 @@ public static void CopyDirectory(string sourceDirName, string destDirName, bool
+            }
+            catch (Exception ex) when (!Debugger.IsAttached)
+            {
+                logger.Error(ex, "Error looking for alternative path");
```
Can you log input path please? It's a last thing, I promise :)

MopsiMauser replied:
ok, it's done ;-)
